### PR TITLE
Classification Submission Form that asks, "Are all these pages done?"

### PR DESCRIPTION
**Pull request by shaunanoordin**
## Feature: Classification Submission Form
- This PR _modifies_ the 'Done' and 'Done & Talk' action with a new intermediary step: before the Classification is submitted to Panoptes, a form asks the user to indicate whether the whole letter has been classified.
- Reason: the question of "are all the lines in all the pages in this letter fully transcribed?" is important because it's the only way of telling whether a letter is ready to be retired or not.

Unfortunately, while we are in agreement that this is important for retirement, the team has not agreed on:
- the question(s) that needs to be asked
- the visual design

### Status
This is currently a WIP. This branch will be rebased once ~#115 is merged~ #113 and #120 are merged.

### CODE REVIEW

**Comment by shaunanoordin** [comment]
EDIT: rebased. Development continues.

**Comment by shaunanoordin** [comment]
## PR Update
- Visually, the Submit Classification Form appears after the 'Done' button is clicked (and prior to the Classification actually being submitted) and appears like this:
screen shot 2017-10-31 at 19 57 34

As you can tell, this is going to need some design magic from @beckyrother . That said, I want this PR to be reviewed for functionality first, and then the UI redesign for the Form can be patched together with updates to #133.

**Comment by shaunanoordin** [comment]
## PR Update
- This PR is now good to go.
- The questions (yes, we can have multiple) and answer can be changed via the PFE Project Builder.
- An example of the Classification (or more specifically, request body) being submitted to `POST /classifications` is...
```
{  
  "http_cache":true,
  "classifications":{  
    "annotations":[  
      {  
        "task":"T1",
        "value":[  
          {  
            "details":[  
              {  
                "value":"Dear Sir,"
              }
            ],
            "points":[  
              {  
                "x":320.59840159840155,
                "y":230.1698301698302
              },
              {  
                "x":404.5144855144855,
                "y":229.3706293706294
              },
              {  
                "x":486.032967032967,
                "y":234.16583416583418
              }
            ],
            "frame":0
          }
        ]
      },
      {  
        "task":"T2",
        "value":"Yes, all pages are complete."
      }
    ],
    "metadata":{  
      "workflow_version":"17.27",
      "started_at":"2017-10-31T20:08:47.579Z",
      "user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.75 Safari/537.36",
      "user_language":"en",
      "utc_offset":"0",
      "subject_dimensions":[  
        {  
          "naturalWidth":638,
          "naturalHeight":800,
          "clientWidth":638,
          "clientHeight":800
        }
      ],
      "finished_at":"2017-10-31T20:09:24.896Z",
      "viewport":{  
        "width":1123,
        "height":932
      }
    },
    "links":{  
      "project":"1764",
      "workflow":"3017",
      "subjects":[  
        "73332"
      ]
    },
    "completed":true
  }
}
```

### Status
Alright, @wgranger , this is ready for review. 👍

**Comment by CKrawczyk** [comment]
Could you update the `value` for the `T2` task to be the *index* of the question taks's answer being picked (e.g. `0` for yes, `1` for no).  That will make sure it is the expected format for a question annotation.

**Comment by shaunanoordin** [comment]
@CKrawczyk can do! In that case I need to make a note to Sam to make sure the order of answers isn't changed, e.g. "1. yes, 2. nope" -> "1. nah, 2. sure thing"

**Comment by CKrawczyk** [comment]
I also just want to confirm if the task label used is taken from the original workflow object?  (The question task is `T1` on production and `T2` on staging)

**Comment by shaunanoordin** [comment]
Update: the Classifications have been updated so that Q&A tasks record the _index_ of the answer, not the answer's _text._

NOTE: The answer value is now a _number,_ not a string. Let me know if anyone can see an issue with this. If there's a problem, I'll just run a .toString() conversion.

Example request payload:
```
{  
  "http_cache":true,
  "classifications":{  
    "annotations":[  
      {  
        "task":"T1",
        "value":[ ... ]
      },
      {  
        "task":"T2",
        "value": 1  //1 for the 'No' answer, 0 for the 'Yes' answer.
      }
    ],
    "metadata":{...}
  }
}
```

Ah! The other issue you just brought up will be hammered out in Slack; I'll follow up with notes + updates on GitHub after I better understand what's needed.

File: src/containers/ClassifierContainer.jsx [hunk 1]
```diff
@@ -49,8 +50,9 @@ class ClassifierContainer extends React.Component {
     this.showCollections = this.showCollections.bind(this);
     this.showTutorial = this.showTutorial.bind(this);
     this.closePopup = this.closePopup.bind(this);
-    this.completeClassification = this.completeClassification.bind(this);
-    this.submitClassificationAndRedirect = this.submitClassificationAndRedirect.bind(this);
+    this.prepareSubmitClassificationForm = this.prepareSubmitClassificationForm.bind(this);
+    //this.completeClassification = this.completeClassification.bind(this);
+    //this.submitClassificationAndRedirect = this.submitClassificationAndRedirect.bind(this);
```
Comment (wgranger) [inline src/containers/ClassifierContainer.jsx]: Why are these merely commented out? Seems like their functionality is replicated in the submit form component.

File: src/containers/ClassifierContainer.jsx [hunk 2]
```diff
@@ -107,10 +109,16 @@ class ClassifierContainer extends React.Component {
...
             <img className="divider" role="presentation" src={Divider} />
+            
+            
+            <button href="#" className="white-green button" onClick={this.prepareSubmitClassificationForm}>Done</button>
+            
+            {/*
             <button href="#" className="white-green button" onClick={this.completeClassification}>Done</button>
             <button href="#" className="green button" onClick={this.submitClassificationAndRedirect}>
               Done &amp; Talk
             </button>
```
Comment (wgranger) [inline src/containers/ClassifierContainer.jsx]: Same here on the commenting out.

Comment (eatyourgreens) [reply src/containers/ClassifierContainer.jsx]: Why have the buttons got href attributes? Those should only appear on links.

**Comment by wgranger** [comment]
I have no conflicts with this branch. Everything appears to be working well, and it seems the T2 answers are kept and then injected into the classification upon clicking the final "done." My only comments are with the commented out code, which is non-blocking.

**Comment by beckyrother** [comment]
The modal also needs a "cancel" button to the left of the "Done" button. Here's a picture of how it could look, using styles from the other modals.

bpl_confirmation modal

**Comment by shaunanoordin** [comment]
Update: 
- Branch updated (via merge, not rebase) to account for the Logging features of PR #135 
  - @wgranger I have some minor queries about the Logging logic, since I'm moving the 'submit classification' triggers from `ClassifierContainer` to `SubmitClassificationForm`
- Commented code is being deleted.
- Cancel button is in the works.

**Comment by shaunanoordin** [comment]
## PR Update
- googleLogger() has been moved to SubmitClassifications. ( @wgranger I'd appreciate a special review look here to ensure data is recorded properly.)
- Old /*commented*/ code has been removed.
- The SubmitClassificationForm now has a Cancel button.
- The SubmitClassificaitonForm has been styled as closely to @beckyrother 's design as I can, but it looks like this:
screen shot 2017-11-02 at 17 40 38
  - Apologies, I wanted to use the same design as the Field Guide (notably, with the curved "top") but it turns out the implementation is different - the Field Guide is a "dialog" component (which doesn't block interaction with items in the screen behind it) while the Submit Classification Form is a Popup component (which does block interaction behind it)
  - The proper solution then is to revise the design of the Popup component (which shares usage with the Tutorial, IIRC) but I'll have to put that into a future ToDo.

### Status
This is now ready for review 👌

**Comment by shaunanoordin** [comment]
OH! @beckyrother I forgot to mention, the first words of the answers couldn't be bolded because the text isn't hardcoded inside the app; the Answer text ("Yes" or "No" or "Yes it's all done" or "No there are still more to be done") is taken from the plaintext provided by Panoptes, so I couldn't just plonk `<b>`s like `<b>Yes</b> it's all done`

(As always, if this is an important design feature though, I can code in an exception that formats on runtime.)

**Comment by shaunanoordin** [comment]
EDIT: forgot to push the latest style branch previously; it should be up now. Be sure to see commit fb6db36: "Updated Done form"

**Comment by wgranger** [comment]
I can still submit this form if I don't select an option. Should we disable "done" and "done and click" if an answer is not provided?

File: src/components/SubmitClassificationForm.jsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+            Done &amp; Talk
+          </button>
+        </div>
+      </div>
+    );
+  }
+  
+  completeClassification() {
+    if (this.context.googleLogger) {
+      this.context.googleLogger.logEvent({ type: 'complete-classification' });
```
Comment (wgranger) [inline src/components/SubmitClassificationForm.jsx]: I can confirm that all googleLogger calls are made correctly in this component.

**Comment by shaunanoordin** [comment]
> I can still submit this form if I don't select an option. Should we disable "done" and "done and click" if an answer is not provided?

That's possible to code in - `submitClassificationForm.render()` can check that the number of answers in matches the number of questions, blocking the Done/Done&Talk when it's invalid. Yeah, I'll post an update shortly-ish

**Comment by shaunanoordin** [comment]
## PR Update
- This PR now prevents users from submitting a Classification unless _all_ questions are answered.

### Status
@wgranger this PR is now ready for an updated review.

**Comment by beckyrother** [comment]
Please remove the shadows from the buttons.

The bold text is not as important, but please add the rest of the sentences ("all pages are complete", "there are still pages to be completed" )

Thanks!

**Comment by shaunanoordin** [comment]
Update:
- shadows from buttons are removed. 👌 
- I'll update the answer text on Panoptes shortly (that's separate from this PR code, but you should still see the data reflected on localhost).

**Comment by wgranger** [comment]
Ok. This looks good to me. The buttons throughout the site are all a bit different, so maybe at some point in the future we can make them look uniform.

**Pull request merged**